src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -366,7 +367,7 @@ private void AddPackageReference(Project project,
         private void AddPackageVersionIntoItemGroupCPM(Project project, LibraryDependency libraryDependency)
         {
             // If onboarded to CPM get the directoryBuildPropsRootElement.
-            ProjectRootElement directoryBuildPropsRootElement = GetDirectoryBuildPropsRootElement(project);
+            ProjectRootElement directoryBuildPropsRootElement = GetDirectoryPackagePropsRootElement(project);
```

Comment:
Can you please help me understand why we're getting a failure that indicates the assets file can't be found in the bug?

I feel like that's a bad error message, and while this may aim to fix the specific scenario bugs, it seems like there could be other cases where we'd get a failure and get a misleading message as well. 
Can we try to fix that too?

Reply:
The error is as follows: `Unable to read a package reference from the project `PATH_TO_CSPROJ_FILE`. Please make sure that your project file and project.assets.json file are in sync by running restore.`

The first sentence could be a little confusing. But I think the second sentence is clear and correct. This error occurs when a library exists in the asset file but could not be found in the project. 

However, we ended up logging this error unexpectedly for the following reasons
* The person had their `GlobalPackageRefernnce` in `Directory.Build.Props` 
* To assert there is no discrepancy between project and lock file, we compare packages between the two, however specifically for `GLobalPackageRefernce` packages, we do it by manually looking at the `Directory.Packages.Props` file. When we could not find the package there, we conclude there is a discrepancy between asset files and the project.

This can be fixed by not manually checking the files but by reading from the project properties instead.

Reply:
> This error occurs when a library exists in the asset file but could not be found in the project.

A similar issue came up recently when a customer was trying to investigate vulnerabilities in their package, so I created this issue: 

* https://github.com/NuGet/Home/issues/13703

I think the ideal solution is that `dotnet list package` shouldn't read any PackageReferences from the project. Only use the MSBuild evaluation to find the assets file, nothing else. Everything else is read from the assets file. This gets rid of any complexity when the assets file and project file are out of sync, so less weird behaviour from the point of view of customers, and less complex code for us. Win-win. Well, except when it's out of sync.

In addition, we should check how we can use the MSBuild evaulation to create a PackageSpec of the project (if we switch MSBuild evaulation to static graph API, we can re-use code from static graph restore.  But I hope it's not too hard to use our existing abstraction to generate a PackageSpec from a `Project` too), so that we can do a restore no-op check (check the PackageSpec hash against the one in `project.nuget.cache`), and when the hash is different, write a warning telling the customer a restore is needed.

Reply:
dotnet list package rendering results from the assets file would definitely be cleaner. However, we defined `dotnet list package` as a command that would list all package references in a project: https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-list-package. And, removing the project properties check would result in us providing "false information" when the two sources of information are out of sync.

However, if we trigger restore every time a user runs `dotnet list package`, we would be able to use the assets file only. This makes it less likely for the assets files and the project to be out of sync. 

I like the idea of using MSBuild evaluation to create a package spec of the project. However, if we have the package spec, do we need to do a restore no-op check? Since the idea is to list the packageRefrences of a project, we could just list the dependencies form the package spec.